pkg/cache/tas_flavor_snapshot.go

```diff
@@ -244,6 +245,66 @@ func (s *TASFlavorSnapshot) removeTASUsage(domainID utiltas.TopologyDomainID, us
 	s.leaves[domainID].tasUsage.Sub(usage)
 }
 
+func (s *TASFlavorSnapshot) freeCapacityPerDomain() map[string]resources.Requests {
+	freeCapacityPerDomain := make(map[string]resources.Requests)
+
+	for domainID, leaf := range s.leaves {
+		freeCapacityPerDomain[string(domainID)] = leaf.freeCapacity.Clone()
+	}
+
+	return freeCapacityPerDomain
+}
+
+func (s *TASFlavorSnapshot) tasUsagePerDomain() map[string]resources.Requests {
+	tasUsagePerDomain := make(map[string]resources.Requests)
+
+	for domainID, leaf := range s.leaves {
+		tasUsagePerDomain[string(domainID)] = leaf.tasUsage.Clone()
+	}
+
+	return tasUsagePerDomain
+}
+
+func (s *TASFlavorSnapshot) PrettyPrintFreeCapacityPerDomain() string {
+	freeCapacityPerDomain := s.freeCapacityPerDomain()
+	tasUsagePerDomain := s.tasUsagePerDomain()
+
+	// Pretty-print the free capacities
+	var prettyDomains []string
+
+	// Sort domain keys lexicographically
+	for _, domain := range slices.Sorted(maps.Keys(freeCapacityPerDomain)) {
+		freeCapacity := freeCapacityPerDomain[domain]
+		tasUsage := tasUsagePerDomain[domain]
+
+		// Pretty-print freeCapacity
+		var freeCapacityDetails []string
+		for _, resourceName := range slices.Sorted(maps.Keys(freeCapacity)) {
+			value := freeCapacity[resourceName]
+			prettyValue := resources.ResourceQuantityString(resourceName, value)
+			freeCapacityDetails = append(freeCapacityDetails, fmt.Sprintf("%s: %s", resourceName, prettyValue))
+		}
+
+		// Pretty-print tasUsage
+		var tasUsageDetails []string
+		for _, resourceName := range slices.Sorted(maps.Keys(tasUsage)) {
+			value := tasUsage[resourceName]
+			prettyValue := resources.ResourceQuantityString(resourceName, value)
+			tasUsageDetails = append(tasUsageDetails, fmt.Sprintf("%s: %s", resourceName, prettyValue))
+		}
+
+		// Format freeCapacity and tasUsage into the output string for this domain
+		prettyDomains = append(prettyDomains, fmt.Sprintf("%s={\nfreeCapacity: %s\ntasUsage: %s\n}",
+			domain,
+			strings.Join(freeCapacityDetails, "; "),
+			strings.Join(tasUsageDetails, "; "),
```

Comment:
```suggestion
			strings.Join(freeCapacityDetails, ", "),
			strings.Join(tasUsageDetails, ", "),
```

I would recommend using JSON like logging form for better log parsing. @mimowo @PBundyra Do you have any motivation for this special log form instead of JSON?

Reply:
I prefer to use JSON, yes, the snipped in https://github.com/kubernetes-sigs/kueue/pull/4295#discussion_r1986072254 lgtm